File: ObjCSample/Tests/ViewControllerAlertTests.m
//  ViewControllerPresentationSpy by Jon Reid, https://qualitycoding.org
//  Copyright 2021 Quality Coding, Inc. See LICENSE.txt

#import "ViewController.h"

@import ViewControllerPresentationSpy;
@import XCTest;

@interface ViewControllerAlertTests : XCTestCase
@end

@implementation ViewControllerAlertTests
{
    QCOAlertVerifier *alertVerifier;
    ViewController *sut;
}

- (void)setUp
{
    [super setUp];
    alertVerifier = [[QCOAlertVerifier alloc] init];
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    sut = [storyboard instantiateViewControllerWithIdentifier:NSStringFromClass([ViewController class])];
    [sut loadViewIfNeeded];
}

- (void)tearDown
{
    alertVerifier = nil;
    sut = nil;
    [super tearDown];
}

- (void)test_outlets_shouldBeConnected
{
    XCTAssertNotNil(sut.showAlertButton);
    XCTAssertNotNil(sut.showActionSheetButton);
}

- (void)test_tappingShowAlertButton_shouldPresentAlert
{
    [sut.showAlertButton sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    XCTAssertEqual(alertVerifier.presentedCount, 1, @"presented count");
    XCTAssertEqual(alertVerifier.preferredStyle, UIAlertControllerStyleAlert, @"preferred style");
    XCTAssertEqual(alertVerifier.presentingViewController, sut, @"presenting view controller");
    XCTAssertTrue(alertVerifier.animated, @"animated");
    XCTAssertEqualObjects(alertVerifier.title, @"Title", @"title");
    XCTAssertEqualObjects(alertVerifier.message, @"Message", @"message");
}

- (void)test_tappingShowActionSheetButton_shouldPresentActionSheet
{
    [sut.showActionSheetButton sendActionsForControlEvents:UIControlEventTouchUpInside];

    XCTAssertEqual(alertVerifier.presentedCount, 1, @"presented count");
    XCTAssertEqual(alertVerifier.preferredStyle, UIAlertControllerStyleActionSheet, @"preferred style");
    XCTAssertEqual(alertVerifier.presentingViewController, sut, @"presenting view controller");
    XCTAssertTrue(alertVerifier.animated, @"animated");
    XCTAssertEqualObjects(alertVerifier.title, @"Title", @"title");
    XCTAssertEqualObjects(alertVerifier.message, @"Message", @"message");
}

- (void)test_popoverForActionSheet
{
    [sut.showActionSheetButton sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    UIPopoverPresentationController *popover = alertVerifier.popover;
    
    XCTAssertEqual(popover.sourceView, sut.showActionSheetButton, @"source view");
    XCTAssertEqualObjects(
            NSStringFromCGRect(popover.sourceRect),
            NSStringFromCGRect(sut.showActionSheetButton.bounds),
            @"source rect");
    XCTAssertEqual(popover.permittedArrowDirections, UIPopoverArrowDirectionAny, @"permitted arrow directions");
}

- (void)test_actionsForAlert
{
    [sut.showAlertButton sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    NSArray<UIAlertAction *> *actions = alertVerifier.actions;
    
    XCTAssertEqual(actions.count, 4);
    XCTAssertEqualObjects(actions[0].title, @"No Handler");
    XCTAssertEqualObjects(actions[1].title, @"Default");
    XCTAssertEqual(actions[1].style, UIAlertActionStyleDefault);
    XCTAssertEqualObjects(actions[2].title, @"Cancel");
    XCTAssertEqual(actions[2].style, UIAlertActionStyleCancel);
    XCTAssertEqualObjects(actions[3].title, @"Destroy");
    XCTAssertEqual(actions[3].style, UIAlertActionStyleDestructive);
}

- (void)test_preferredActionForAlert
{
    [sut.showAlertButton sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    XCTAssertEqualObjects(alertVerifier.preferredAction.title, @"Default");
}

- (void)test_actionsForActionSheet
{
    [sut.showActionSheetButton sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    NSArray<UIAlertAction *> *actions = alertVerifier.actions;
    
    XCTAssertEqual(actions.count, 4);
    XCTAssertEqualObjects(actions[1].title, @"Default");
    XCTAssertEqual(actions[1].style, UIAlertActionStyleDefault);
    XCTAssertEqualObjects(actions[2].title, @"Cancel");
    XCTAssertEqual(actions[2].style, UIAlertActionStyleCancel);
    XCTAssertEqualObjects(actions[3].title, @"Destroy");
    XCTAssertEqual(actions[3].style, UIAlertActionStyleDestructive);
}

- (void)test_executeActionForButtonWithTitle_withDefaultButton_shouldExecuteDefaultAction
{
    [sut.showAlertButton sendActionsForControlEvents:UIControlEventTouchUpInside];

    NSError *error = nil;
    [alertVerifier executeActionForButton:@"Default" andReturnError:&error];

    XCTAssertNil(error);
    XCTAssertEqual(sut.alertDefaultActionCount, 1);
}

- (void)test_executeActionForButtonWithTitle_withCancelButton_shouldExecuteCancelAction
{
    [sut.showAlertButton sendActionsForControlEvents:UIControlEventTouchUpInside];

    NSError *error = nil;
    [alertVerifier executeActionForButton:@"Cancel" andReturnError:&error];

    XCTAssertNil(error);
    XCTAssertEqual(sut.alertCancelActionCount, 1);
}

- (void)test_executeActionForButtonWithTitle_withDestroyButton_shouldExecuteDestroyAction
{
    [sut.showAlertButton sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    NSError *error = nil;
    [alertVerifier executeActionForButton:@"Destroy" andReturnError:&error];

    XCTAssertNil(error);
    XCTAssertEqual(sut.alertDestroyActionCount, 1);
}

- (void)test_textFieldsForAlert
{
    [sut.showAlertButton sendActionsForControlEvents:UIControlEventTouchUpInside];

    XCTAssertEqual(alertVerifier.textFields.count, 1);
    XCTAssertEqualObjects(alertVerifier.textFields[0].placeholder, @"Placeholder");
}

- (void)test_textFields_shouldNotBeAddedToActionSheets
{
    [sut.showActionSheetButton sendActionsForControlEvents:UIControlEventTouchUpInside];

    XCTAssertEqual(alertVerifier.textFields.count, 0);
}

@end

File: SwiftSample/Tests/ViewControllerAlertTests.swift
// ViewControllerPresentationSpy by Jon Reid, https://qualitycoding.org
// Copyright 2015 Jonathan M. Reid. https://github.com/jonreid/ViewControllerPresentationSpy/blob/main/LICENSE.txt
// SPDX-License-Identifier: MIT

@testable import SwiftSampleViewControllerPresentationSpy
import ViewControllerPresentationSpy
import XCTest

@MainActor
final class ViewControllerAlertTests: XCTestCase, Sendable {
    private var alertVerifier: AlertVerifier!
    private var sut: ViewController!

    override func setUp() async throws {
        try await super.setUp()
        alertVerifier = AlertVerifier()
        let storyboard = UIStoryboard(name: "Main", bundle: nil)
        sut = storyboard.instantiateViewController(identifier: String(describing: ViewController.self))
        sut.loadViewIfNeeded()
    }

    override func tearDown() async throws {
        alertVerifier = nil
        sut = nil
        try await super.tearDown()
    }

    func test_outlets_areConnected() throws {
        XCTAssertNotNil(sut.showAlertButton)
        XCTAssertNotNil(sut.showActionSheetButton)
    }

    func test_tappingShowAlertButton_showsAlert() throws {
        sut.showAlertButton.sendActions(for: .touchUpInside)

        alertVerifier.verify(
            title: "Title",
            message: "Message",
            animated: true,
            actions: [
                .default("No Handler"),
                .default("Default"),
                .cancel("Cancel"),
                .destructive("Destroy"),
            ]
        )
    }

    func test_tappingShowActionSheetButton_showsActionSheet() throws {
        sut.showActionSheetButton.sendActions(for: .touchUpInside)

        alertVerifier.verify(
            title: "Title",
            message: "Message",
            animated: true,
            actions: [
                .default("No Handler"),
                .default("Default"),
                .cancel("Cancel"),
                .destructive("Destroy"),
            ],
            preferredStyle: .actionSheet
        )
    }

    func test_popoverForActionSheet() throws {
        sut.showActionSheetButton.sendActions(for: .touchUpInside)

        let popover = try XCTUnwrap(alertVerifier.popover)

        XCTAssertEqual(popover.sourceView, sut.showActionSheetButton, "source view")
        XCTAssertEqual("\(popover.sourceRect)", "\(sut.showActionSheetButton.bounds)", "source rect")
        XCTAssertEqual(popover.permittedArrowDirections, UIPopoverArrowDirection.any, "permitted arrow directions")
    }

    func test_preferredActionForAlert() throws {
        sut.showAlertButton.sendActions(for: .touchUpInside)

        XCTAssertEqual(alertVerifier.preferredAction?.title, "Default")
    }

    func test_executeActionForButton_withDefaultButton_executesDefaultAction() throws {
        sut.showAlertButton.sendActions(for: .touchUpInside)

        try alertVerifier.executeAction(forButton: "Default")

        XCTAssertEqual(sut.alertDefaultActionCount, 1)
    }

    func test_executeActionForButton_withCancelButton_executesCancelAction() throws {
        sut.showAlertButton.sendActions(for: .touchUpInside)

        try alertVerifier.executeAction(forButton: "Cancel")

        XCTAssertEqual(sut.alertCancelActionCount, 1)
    }

    func test_executeActionForButton_withDestroyButton_executesDestroyAction() throws {
        sut.showAlertButton.sendActions(for: .touchUpInside)

        try alertVerifier.executeAction(forButton: "Destroy")

        XCTAssertEqual(sut.alertDestroyActionCount, 1)
    }

    func test_textFieldsForAlert() throws {
        sut.showAlertButton.sendActions(for: .touchUpInside)

        XCTAssertEqual(alertVerifier.textFields?.count, 1)
        XCTAssertEqual(alertVerifier.textFields?[0].placeholder, "Placeholder")
    }

    func test_textFieldsAreNotAddedToActionSheets() throws {
        sut.showActionSheetButton.sendActions(for: .touchUpInside)

        XCTAssertEqual(alertVerifier.textFields?.count, 0)
    }
}
